File: src/main/java/org/kite9/diagram/batik/element/TerminatorImpl.java
package org.kite9.diagram.batik.element;

import org.apache.batik.anim.dom.SVGOMMarkerElement;
import org.apache.batik.css.engine.value.Value;
import org.apache.batik.css.engine.value.ValueConstants;
import org.kite9.diagram.batik.bridge.Kite9BridgeContext;
import org.kite9.diagram.batik.bridge.Kite9DocumentLoader;
import org.kite9.diagram.batik.bridge.RectangularPainter;
import org.kite9.diagram.model.Connection;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.Leaf;
import org.kite9.diagram.model.Terminator;
import org.kite9.diagram.model.position.CostedDimension;
import org.kite9.diagram.model.position.Dimension2D;
import org.kite9.diagram.model.position.Direction;
import org.kite9.framework.common.Kite9ProcessingException;
import org.kite9.framework.dom.CSSConstants;
import org.kite9.framework.logging.LogicException;
import org.kite9.framework.xml.StyledKite9SVGElement;

public class TerminatorImpl extends AbstractRectangularDiagramElement implements Terminator {
	
	private SVGOMMarkerElement markerElement;
	private Value reference;

	public TerminatorImpl(StyledKite9SVGElement el, DiagramElement parent, Kite9BridgeContext ctx, RectangularPainter<Leaf> rp) {
		super(el, parent, ctx, rp);
	}

	@Override
	protected void initialize() {
		reference = theElement.getCSSStyleProperty(CSSConstants.MARKER_REFERENCE);
		if (reference != ValueConstants.NONE_VALUE) {
			Kite9DocumentLoader loader = (Kite9DocumentLoader) ctx.getDocumentLoader();
			markerElement = (SVGOMMarkerElement) loader.loadElementFromUrl(reference, theElement);
			if (markerElement != null) {
				double[] elemSizes = getSizesFromElement(markerElement);
				for (int i = 0; i < padding.length; i++) {
					padding[i] = Math.max(padding[i], elemSizes[i]);
				}
			}	
		} 
	}
	
	private double[] getSizesFromElement(SVGOMMarkerElement markerElement) {
		double markerWidth=markerElement.getMarkerWidth().getAnimVal().getValue();
		double markerHeight=markerElement.getMarkerHeight().getAnimVal().getValue();
		double refX=markerElement.getRefX().getAnimVal().getValue();
		double refY=markerElement.getRefY().getAnimVal().getValue();
		return new double[] {
			refY, 
			markerWidth - refX,
			markerHeight - refY,
			refX
		};
	}

	@Override
	public Container getContainer() {
		Connection c = (Connection) getParent();
		if (this == c.getFromDecoration()) {
			return c.getFrom().getContainer();
		} else if (this == c.getToDecoration()) {
			return c.getTo().getContainer();
		} else {
			throw new Kite9ProcessingException();
		}
	}

	@Override
	public double getReservedLength() {
		ensureInitialized();
		return getPadding(Direction.LEFT) + getMargin(Direction.LEFT);
	}

	public String getMarkerUrl() {
		ensureInitialized();
		if (markerElement != null) {
			return "url(#"+markerElement.getId()+")";
		} else {
			return null;
		}
	}

	@Override
	public CostedDimension getSize(Dimension2D within) {
		throw new LogicException("Shouldn't be using size for terminators");
	}

}

File: kite9-visualization/src/commonMain/kotlin/org/kite9/diagram/dom/model/TerminatorImpl.kt
package org.kite9.diagram.dom.model

import org.kite9.diagram.dom.bridge.ElementContext
import org.kite9.diagram.dom.css.CSSConstants
import org.kite9.diagram.dom.painter.Painter
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.Connection
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.Rectangular
import org.kite9.diagram.model.Temporary
import org.kite9.diagram.model.Terminator
import org.kite9.diagram.model.position.CostedDimension2D.Companion.ZERO
import org.kite9.diagram.model.position.Dimension2D
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.End
import org.kite9.diagram.model.style.BorderTraversal
import org.kite9.diagram.model.style.ContentTransform
import org.kite9.diagram.model.style.DiagramElementSizing
import org.w3c.dom.Element

class TerminatorImpl(
    el: Element,
    parent: DiagramElement,
    ctx: ElementContext,
    rp: Painter,
    t: ContentTransform,
    private val end: End
) : AbstractRectangular(
    el, parent, ctx, rp, t
), Terminator {

    private var markerReserve = 0.0
    private var arrivalSide: Direction? = null

    override fun initialize() {
        super.initialize()
        arrivalSide = ElementContext.getCssStyleEnumProperty<Direction>(CSSConstants.DIRECTION, theElement, ctx)
        markerReserve = ctx.getCssStyleDoubleProperty(CSSConstants.MARKER_RESERVE, theElement)
    }

    override fun initContainerPosition() {

    }

    override fun getContainer(): Rectangular? {
        val c = getConnection()
        return when {
            this === c.getFromDecoration() -> {
                c.getFrom().getContainer()
            }
            this === c.getToDecoration() -> {
                c.getTo().getContainer()
            }
            else -> {
                 throw ctx.contextualException("Couldn't get container for terminator " + getID(), theElement)
            }
        }
    }

    override fun getReservedLength(): Double {
        ensureInitialized()
        return getPadding(Direction.RIGHT) + getMargin(Direction.RIGHT)
    }

    override fun getMarkerReserve(): Double {
        ensureInitialized()
        return markerReserve
    }

    override fun styleMatches(t2: Terminator?): Boolean {
        return if (t2 is TerminatorImpl) {
            val styleMatch = attributesMatch("style", this, t2)
            val classMatch = attributesMatch("class", this, t2)
            styleMatch && classMatch
        } else {
            false
        }
    }

    override fun getMinimumSize(): Dimension2D {
        return ZERO
    }

    override fun getXPathVariable(name: String): String? {
        if ("x0" == name || "y0" == name) {
            return "0"
        } else if ("x1" == name || "width" == name) {
            return "" + getMinimumSize().width()
        } else if ("y1" == name || "height" == name) {
            return "" + getMinimumSize().height()
        }
        return null
    }

    override fun getArrivalSide(): Direction? {
        ensureInitialized()
        return arrivalSide
    }

    override fun getConnection(): Connection {
        ensureInitialized()
        var parent = getParent()
        while ((parent !is Connection) && (parent != null)) {
            parent = getParent()
        }
        if (parent is Connection) {
            return parent
        } else {
            throw ctx.contextualException("Parent is not link for terminator", theElement)
        }
    }

    override fun getEnd(): End {
        return end
    }

    override fun getSizing(horiz: Boolean): DiagramElementSizing {
        return DiagramElementSizing.MINIMIZE
    }

    override fun getContents(): List<DiagramElement> {
        return emptyList()
    }

    override fun addTemporaryContent(t: Temporary) {
        throw LogicException("Can't add contents to terminators")
    }

    override fun getTraversalRule(d: Direction): BorderTraversal {
        throw LogicException("Can't traverse terminators")
    }

    override fun getGridColumns(): Int {
        return 0
    }

    override fun getGridRows(): Int {
        return 0
    }

    companion object {
        private fun attributesMatch(name: String, a: TerminatorImpl, b: TerminatorImpl): Boolean {
            return a.theElement.getAttribute(name) ==
                    b.theElement.getAttribute(name)
        }
    }
}